pkg/build/builder/common_test.go

```diff
@@ -113,3 +121,381 @@ func TestContainerName(t *testing.T) {
+				FROM scratch as test
+				COPY --from=test /a /b
+				`),
+			build: []buildapiv1.ImageSource{
+				{From: corev1.ObjectReference{Kind: "DockerImage", Name: "nginx:latest"}, As: []string{"test", "scratch"}},
+			},
+			want: want{
+				Out: heredoc.Doc(`
+				FROM nginx:latest
+				FROM nginx:latest as test
+				COPY --from=test /a /b
+				`),
+			},
+		},
+	}
+	for i, test := range tests {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			f, err := ioutil.TempFile("", "builder-dockertest")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer os.Remove(f.Name())
+			defer f.Close()
+			if _, err := f.Write([]byte(test.original)); err != nil {
+				t.Fatal(err)
+			}
+			f.Close()
+			if _, err := dockerfile.Parse(strings.NewReader(test.original)); err != nil {
+				t.Fatal(err)
+			}
+			if _, err := dockerfile.Parse(strings.NewReader(test.want.Out)); err != nil {
+				t.Fatal(err)
+			}
+			build := &buildapiv1.Build{}
+			build.Spec.Strategy.DockerStrategy = &buildapiv1.DockerBuildStrategy{
+				DockerfilePath: filepath.Base(f.Name()),
+			}
+			if test.from != nil {
+				build.Spec.Strategy.DockerStrategy.From = test.from
+			}
+			build.Spec.Source.Images = test.build
+			sourceInfo := &git.SourceInfo{}
+			testErr := addBuildParameters(filepath.Dir(f.Name()), build, sourceInfo)
+			out, err := ioutil.ReadFile(f.Name())
+			if err != nil {
+				t.Fatal(err)
+			}
+			got := want{
+				Err: testErr != nil,
+				Out: string(out),
+			}
+			extra := "ENV \"OPENSHIFT_BUILD_NAME\"=\"\" \"OPENSHIFT_BUILD_NAMESPACE\"=\"\"\nLABEL \"io.openshift.build.name\"=\"\" \"io.openshift.build.namespace\"=\"\"\n"
+			test.want.Out += extra
+			if !reflect.DeepEqual(test.want, got) {
+				t.Errorf("unexpected: %s", diff.ObjectReflectDiff(test.want, got))
+			}
+		})
+	}
+}
+
+func Test_findReferencedImages(t *testing.T) {
+	type want struct {
+		Images     []string
+		Multistage bool
+		Err        bool
+	}
+	tests := []struct {
+		original string
+		want     want
+	}{
+		{
+			original: `# no FROM instruction`,
+			want: want{
+				Images: []string{},
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch
+				# FROM busybox
+				RUN echo "hello world"
+				`),
+			want: want{
+				Images: []string{"scratch"},
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch
+				FROM busybox
+				RUN echo "hello world"
+				`),
+			want: want{
+				Images:     []string{"busybox", "scratch"},
+				Multistage: true,
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				FROM busybox
+				RUN echo "hello world"
+				`),
+			want: want{
+				Images:     []string{"busybox", "scratch"},
+				Multistage: true,
+			},
+		},
+		{
+			original: heredoc.Doc(`
+				FROM scratch as test
+				FROM busybox
+				COPY --from=test /a /b
+				COPY --from=nginx /a /c
+				COPY --from=nginx:latest /a /c
+				RUN echo "hello world"
+				`),
+			want: want{
+				Images:     []string{"busybox", "nginx", "nginx:latest", "scratch"},
```

Comment:
same thing here...test for when the `--from=y` reference comes before the `FROM x as y` definition.